tests/analog/test_patterns.py

```diff
@@ -0,0 +1,125 @@
+from __future__ import annotations
+
+import pytest
+import torch
+from metrics import ATOL_DICT
+
+from qadence import (
+    AnalogRX,
+    BackendName,
+    DifferentiableBackend,
+    DiffMode,
+    Parameter,
+    QuantumCircuit,
+    QuantumModel,
+    total_magnetization,
+)
+from qadence.analog.addressing import AddressingPattern
+from qadence.analog.interaction import add_interaction
+from qadence.backends.pulser.backend import Backend as PulserBackend
+from qadence.backends.pulser.config import Configuration
+from qadence.backends.pyqtorch.backend import Backend as PyqBackend
+
+
+@pytest.mark.parametrize(
+    "max_amp,max_det",
+    [(0.0, 10.0), (15.0, 0.0), (15.0, 9.0)],
+)
+@pytest.mark.parametrize(
+    "spacing",
+    [8.0, 30.0],
+)
+def test_pulser_pyq_addressing(max_amp: float, max_det: float, spacing: float) -> None:
+    n_qubits = 3
+    block = AnalogRX("x")
+    circ = QuantumCircuit(n_qubits, block)
+
+    # define addressing patterns
+    rand_weights_amp = torch.rand(n_qubits)
+    rand_weights_amp = rand_weights_amp / rand_weights_amp.sum()
+    w_amp = {i: rand_weights_amp[i] for i in range(n_qubits)}
+    rand_weights_det = torch.rand(n_qubits)
+    rand_weights_det = rand_weights_det / rand_weights_det.sum()
+    w_det = {i: rand_weights_det[i] for i in range(n_qubits)}
+    p = AddressingPattern(
+        n_qubits=n_qubits,
+        max_det=max_det,
+        max_amp=max_amp,
+        weights_det=w_det,
+        weights_amp=w_amp,
+    )
+
+    values = {"x": torch.linspace(0.5, 2 * torch.pi, 50)}
+    obs = total_magnetization(n_qubits)
+    conf = Configuration(addressing_pattern=p, spacing=spacing)
+
+    # define pulser backend
+    pulser_backend = PulserBackend(config=conf)  # type: ignore[arg-type]
+    conv = pulser_backend.convert(circ, obs)
+    pulser_circ, pulser_obs, embedding_fn, params = conv
+    diff_backend = DifferentiableBackend(pulser_backend, diff_mode=DiffMode.GPSR)
+    expval_pulser = diff_backend.expectation(pulser_circ, pulser_obs, embedding_fn(params, values))
+
+    # define pyq backend
+    int_circ = add_interaction(circ, spacing=spacing, pattern=p)
+    pyq_backend = PyqBackend()  # type: ignore[arg-type]
```

Comment:
This kind of breaks the interface for the `Analog` blocks, right? AFAIK as of today the interaction is added automatically for the analog blocks in a transpilation pass for the PyQ backend. I guess this is temporary while @jpmoutinho is revising the whole structure for the emulated analog interface, is that right?

Reply:
Indeed the usage of `add_interaction` here is temporary. You can check the `add_background_hamiltonian` function in https://github.com/pasqal-io/qadence/pull/176/ (parse_analog.py) which is where @vytautas-a and me envisioned the addressing pattern will be handled. That function replaces `add_interaction` and was renamed given that it will no longer be just the interaction. Then it is still handled automatically by the transpilation pass in the PyQ backend which will get the addressing pattern from the configuration the user makes in the device.

Reply:
It's added in transpilation pass, yes, but you can call this function again externally if you want to set spacing or, in this case, pass the pattern.